### PR TITLE
Add model complexity badge

### PR DESCRIPTION
**Pull request by MoritzWeber0**

### CODE REVIEW

**Review by amolenaar — CHANGES_REQUESTED** [review]
All-in-all pretty okay. Only I dislike the workaround to avoid triggering the error interceptor in the frontend. There must be a better way.

File: backend/capellacollab/projects/toolmodels/modelbadge/routes.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                    "reason": (
+                        "No model complexity badge found in the linked Git repository.",
+                        "Please contact your administrator.",
+                    ),
+                },
+            )
+
+        logger.info("Failed fetching model complexity badge", exc_info=True)
+        raise fastapi.HTTPException(
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/projects/toolmodels/modelbadge/routes.py:66]: Why? The 404 (not found) case is catched before. This is the case where something unknown happens (HTTPError with status code other than 404 from Gitlab).

File: frontend/src/app/general/error-handling/error-handling.interceptor.ts [hunk 2]
```diff
@@ -30,6 +30,9 @@ export class ErrorHandlingInterceptor implements HttpInterceptor {
       catchError(this.constructErrorDetailTransformationObservable),
       tap({
         next: (event: HttpEvent<any>) => {
+          if (request.headers.get('skip-frontend-error-handling') == 'true') {
+            return;
+          }
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply frontend/src/app/general/error-handling/error-handling.interceptor.ts]: I found a way to use headers instead: https://github.com/DSD-DBS/capella-collab-manager/pull/665/commits/a91a5d0d2e5653c04a53199ba9d9502785df9875

File: frontend/src/app/projects/project-detail/model-overview/model-complexity-badge/service/model-complexity-badge.service.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    modelSlug: string
+  ): Observable<Blob> {
+    return this.http.get(
+      this.modelService.backendURLFactory(projectSlug, modelSlug) +
+        '/badges/complexity',
+      {
+        responseType: 'blob',
+        headers: {
+          'Skip-Frontend-Error-Handling': 'true',
+        },
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply frontend/src/app/projects/project-detail/model-overview/model-complexity-badge/service/model-complexity-badge.service.ts]: Status codes don't help here - I want to disable the automated error handling complexity (no matter which status code). Usually, a toast message appears, but in this case, I found it pretty annoying, so the error message is printed in the context diagram box instead.

File: docs/user/docs/projects/models/complexity_badge.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+1. Follow the
+   [Gitlab CI template instructions](https://github.com/DSD-DBS/py-capellambse/tree/master/ci-templates/gitlab#model-badge)
+   to add the complexity badge pipeline to the `.gitlab-ci.yml` of the linked
+   repository.
+
+## View the model complexity badge
+
+The model complexity badge is visible in the model overview in the project
+details.
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply docs/user/docs/projects/models/complexity_badge.md]: Done.

File: backend/capellacollab/projects/toolmodels/modelbadge/routes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/projects/toolmodels/modelbadge/routes.py:66]: It's definitely a "weak" point of the backend - as we can't really support the handling of all error cases (there are just too many). So it's also a server error - we're just not handle the "unknown" error. 

On the other hand, it's counter-intuitive as you mentioned. 

One possible middle ground would be to pass all status codes of Gitlab and use them as "our" status code too.

[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/projects/toolmodels/modelbadge/routes.py:66]: Agreed with @amolenaar to handle it as separate issue.

**Comment by MoritzWeber0** [comment]
I've added test coverage in https://github.com/DSD-DBS/capella-collab-manager/pull/665/commits/2b0d9dbfbc32133ae8a5e88c573b1eea476c2ebd. 
Would be great if you can review the new commit @amolenaar.

File: backend/tests/toolmodels/test_model_badge.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+            <rect fill="#91cc84" stroke="#333" stroke-width="0" x="37" y="25" width="5" height="3" />
+            <text x="43" y="27.6">System Analysis</text>
+            <rect fill="#a5c2e6" stroke="#333" stroke-width="0" x="67" y="25" width="5" height="3" />
+            <text x="73" y="27.6">Logical Architecture</text>
+            <rect fill="#f89f9f" stroke="#333" stroke-width="0" x="100" y="25" width="5" height="3" />
+            <text x="106" y="27.6">Physical Architecture</text>
+        </g>
+    </g>
+</svg>
+"""
```
Comment (amolenaar) [inline backend/tests/toolmodels/test_model_badge.py:63]: I suppose a simple placeholder text would do, e.g. `<svg>badge placeholder</svg>`. This is a bit overwhelming ;).

**Review by amolenaar — APPROVED** [review]
Looks good.

I'm not familiar with the `@responses.activate` decorator, but the solution looks neat.

Is there a reason all fixtures start with `fixture_` (something to do with linters?), otherwise I would use the function name as fixture name.

File: backend/tests/toolmodels/test_model_badge.py (hunk 5 above)
Comment (MoritzWeber0) [reply backend/tests/toolmodels/test_model_badge.py]: Done.

**Comment by MoritzWeber0** [comment]
> Is there a reason all fixtures start with `fixture_` (something to do with linters?), otherwise I would use the function name as fixture name.

Yes, there is. It makes pylint happy :) 
Otherwise, it complains about redefined-outer-name when you want to use it as argument.

**Pull request merged**